src/controller/ReadInteraction.h

```diff
@@ -25,30 +25,24 @@
 
 namespace chip {
 namespace Controller {
+namespace detail {
 
-/*
- * A typed read attribute function that takes as input a template parameter that encapsulates the type information
- * for a given attribute as well as callbacks for success and failure and either returns a decoded cluster-object representation
- * of the requested attribute through the provided success callback or calls the provided failure callback.
- *
- * The AttributeTypeInfo is generally expected to be a ClusterName::Attributes::AttributeName::TypeInfo struct, but any
- * object that contains type information exposed through a 'DecodableType' type declaration as well as GetClusterId() and
- * GetAttributeId() methods is expected to work.
- *
- */
+template <typename DecodableAttributeType>
+struct ReportParams : public app::ReadPrepareParams
+{
+    ReportParams(SessionHandle sessionHandle) : app::ReadPrepareParams(sessionHandle) { mKeepSubscriptions = false; }
+    typename TypedReadAttributeCallback<DecodableAttributeType>::OnSuccessCallbackType mOnReportCb;
+    typename TypedReadAttributeCallback<DecodableAttributeType>::OnErrorCallbackType mOnErrorCb;
+    typename TypedReadAttributeCallback<DecodableAttributeType>::OnSubscriptionEstablishedCallbackType
+        mOnSubscriptionEstablishedCb             = nullptr;
+    app::ReadClient::InteractionType mReportType = app::ReadClient::InteractionType::Read;
+};
 
-/**
- * To avoid instantiating all the complicated code on a per-attribute basis, we
- * have a helper that's just templated on the type.
- */
 template <typename DecodableAttributeType>
-CHIP_ERROR ReadAttribute(Messaging::ExchangeManager * aExchangeMgr, const SessionHandle sessionHandle, EndpointId endpointId,
-                         ClusterId clusterId, AttributeId attributeId,
-                         typename TypedReadAttributeCallback<DecodableAttributeType>::OnSuccessCallbackType onSuccessCb,
-                         typename TypedReadAttributeCallback<DecodableAttributeType>::OnErrorCallbackType onErrorCb)
+CHIP_ERROR ReportAttribute(Messaging::ExchangeManager * exchangeMgr, EndpointId endpointId, ClusterId clusterId,
+                           AttributeId attributeId, ReportParams<DecodableAttributeType> && readParams)
```

Comment:
Should also update the ReadEvent as well?

Reply:
I'd like to keep these PRs somewhat scoped: this is needed to replace existing broken code; the ReadEvent case would be new code that importantly would not be exercised by anything now and so hard to test.  I'm happy to do it, but in a followup once we have use cases for it.  Assuming they appear.